include/ginkgo/extensions/kokkos/types.hpp

```diff
@@ -0,0 +1,288 @@
+                  "type with mismatching type sizes.");
+    static_assert(
+        alignof(std::decay_t<T>) == alignof(std::decay_t<type>),
+        "Can't handle C++ data type and corresponding Kokkos type with "
+        "mismatching alignments. If std::complex is used, please make sure "
+        "to configure Kokkos with `KOKKOS_ENABLE_COMPLEX_ALIGN=ON`");
+};
+
+template <typename T>
+using value_type_t = typename value_type<T>::type;
+
+
+template <typename T, typename MemorySpace>
+struct mapper {
+    static auto map(T&);
+    static auto map(const T&);
+};
+
+/**
+ * Type that maps a Ginkgo array to an unmanaged 1D Kokkos::View.
+ *
+ * @warning Using std::complex as data type might lead to issues, since the
+ *          alignment of Kokkos::complex is not necessarily the same.
+ *
+ * @tparam MemorySpace  The memory space type the mapped object should use.
+ */
+template <typename ValueType, typename MemorySpace>
+struct mapper<array<ValueType>, MemorySpace> {
+    template <typename ValueType_c>
+    using type =
+        Kokkos::View<typename value_type<ValueType_c>::type*, MemorySpace,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+    template <typename ValueType_c>
+    static type<ValueType_c> map(ValueType_c* data, size_type size)
+    {
+        return type<ValueType_c>{
+            reinterpret_cast<value_type_t<ValueType_c>*>(data), size};
+    }
+
+    static type<ValueType> map(array<ValueType>& arr)
+    {
+        assert_compatibility(arr, MemorySpace{});
+
+        return map(arr.get_data(), arr.get_size());
+    }
+
+    static type<const ValueType> map(const array<ValueType>& arr)
+    {
+        assert_compatibility(arr, MemorySpace{});
+
+        return map(arr.get_const_data(), arr.get_size());
+    }
+
+
+    static type<const ValueType> map(
+        const ::gko::detail::const_array_view<ValueType>& arr)
+    {
+        assert_compatibility(arr, MemorySpace{});
+
+        return map(arr.get_const_data(), arr.get_size());
+    }
+};
+
+
+/**
+ * Type that maps a Ginkgo matrix::Dense to an unmanaged 2D Kokkos::View.
+ *
+ * @warning Using std::complex as data type might lead to issues, since the
+ *          alignment of Kokkos::complex is not necessarily the same.
+ *
+ * @tparam MemorySpace  The memory space type the mapped object should use.
+ */
+template <typename ValueType, typename MemorySpace>
+struct mapper<matrix::Dense<ValueType>, MemorySpace> {
+    template <typename ValueType_c>
+    using type = Kokkos::View<typename value_type<ValueType_c>::type**,
+                              Kokkos::LayoutStride, MemorySpace,
+                              Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+    static type<ValueType> map(matrix::Dense<ValueType>& m)
+    {
+        assert_compatibility(m, MemorySpace{});
+
+        auto size = m.get_size();
+
+        return type<ValueType>{
+            reinterpret_cast<value_type_t<ValueType>*>(m.get_values()),
```

Comment:
will it be kind of undefined behavior?

Reply:
It's doing the same as our `as_device_type` functions, so from that perspective it should be fine. But TBH I'm not sure if both are undefined behavior anyway.